src/main/java/mate/academy/AsyncRequestProcessor.java

```diff
@@ -1,16 +1,28 @@
 package mate.academy;
 
+import java.util.Map;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Executor;
 
 public class AsyncRequestProcessor {
     private final Executor executor;
+    private final Map<String, UserData> cache = new ConcurrentHashMap<>();
 
     public AsyncRequestProcessor(Executor executor) {
         this.executor = executor;
     }
 
     public CompletableFuture<UserData> processRequest(String userId) {
-        return null;
+        UserData cachedData = cache.get(userId);
+        if (cachedData != null) {
+            return CompletableFuture.completedFuture(cachedData);
+        }
+        return CompletableFuture
+                .supplyAsync(() -> new UserData(userId, "Details for " + userId), executor)
+                .thenApply(userData -> {
+                    cache.putIfAbsent(userId, userData);
```

Comment:
The use of 'cache.putIfAbsent' is correct for thread-safe caching, but since the retrieval and the computation of the value are not atomic, consider using 'cache.computeIfAbsent' to ensure that the function to compute the value is applied at most once per key.